src/integrationTest/java/uk/gov/hmcts/reform/blobrouter/services/NotificationServiceTest.java

```diff
@@ -0,0 +1,138 @@
+        notificationService = new NotificationService(
+            notificationsPublisher,
+            rejectedEnvelopeRepository,
+            envelopeService
+        );
+    }
+
+    @AfterEach
+    void tearDown() {
+        dbHelper.deleteAll();
+    }
+
+    @Test
+    void should_send_notifications_for_the_rejected_envelopes() {
+        // given
+        // rejected envelope
+        var envelopeId1 = envelopeService.createNewEnvelope("bulkscan", "blob1.zip", now());
+        envelopeService.markAsRejected(envelopeId1, "duplicate file");
+
+        // envelope that is not rejected
+        var envelopeId2 = envelopeService.createNewEnvelope("bulkscan", "blob2.zip", now());
+        envelopeService.markAsDispatched(envelopeId2);
+
+        // rejected envelope
+        var envelopeId3 = envelopeService.createNewEnvelope("testcontainer", "blob3.zip", now());
+        envelopeService.markAsRejected(envelopeId3, "invalid signature");
+        envelopeService.saveEvent(envelopeId3, DELETED);
+
+        // when
+        notificationService.sendNotifications();
+
+        // then
+        verify(notificationsPublisher, times(2)).publish(any());
+
+        Optional<Envelope> envelope1 = envelopeService.findEnvelope(envelopeId1);
+        assertThat(envelope1).hasValueSatisfying(env -> assertThat(env.pendingNotification).isFalse());
+        List<EnvelopeEvent> envelope1Events = envelopeEventRepository.findForEnvelope(envelopeId1);
+        assertThat(envelope1Events).extracting(e -> e.type).contains(EventType.NOTIFICATION_SENT);
+
+        Optional<Envelope> envelope3 = envelopeService.findEnvelope(envelopeId3);
+        assertThat(envelope3).hasValueSatisfying(env -> assertThat(env.pendingNotification).isFalse());
+        List<EnvelopeEvent> envelope3Events = envelopeEventRepository.findForEnvelope(envelopeId3);
+        assertThat(envelope3Events).extracting(e -> e.type).contains(EventType.NOTIFICATION_SENT);
+    }
+
+    @Test
+    void should_not_send_notification_when_notification_was_already_sent() {
+        // given
+        // rejected and notification_sent
+        var envelopeId1 = envelopeService.createNewEnvelope("bulkscan", "blob1.zip", now());
+        envelopeService.markAsRejected(envelopeId1, "duplicate file");
+        envelopeService.saveEvent(envelopeId1, DELETED);
+        envelopeService.markPendingNotificationAsSent(envelopeId1);
+
+        // envelope that is not rejected
+        var envelopeId2 = envelopeService.createNewEnvelope("bulkscan", "blob2.zip", now());
+        envelopeService.markAsDispatched(envelopeId2);
+        envelopeService.saveEvent(envelopeId1, DELETED);
+
+        // rejected but notification NOT sent
+        var envelopeId3 = envelopeService.createNewEnvelope("bulkscan", "blob3.zip", now());
+        envelopeService.markAsRejected(envelopeId3, "invalid signature");
+        envelopeService.saveEvent(envelopeId3, DELETED);
+
+        // when
+        notificationService.sendNotifications();
```

Comment:
how about we call the service twice?
first time it should send notifications, the second time it shouldn't